packages/fxa-auth-server/lib/payments/stripe.ts

```diff
@@ -3120,7 +3105,40 @@ export class StripeHelper extends StripeHelperBase {
         ? SUBSCRIPTION_UPDATE_TYPES.UPGRADE
         : SUBSCRIPTION_UPDATE_TYPES.DOWNGRADE;
 
-    const productPaymentCycleOld = this.stripePlanToPaymentCycle(planOld);
+    const productPaymentCycleOld = await this.stripePlanToPaymentCycle(planOld);
+
+    const sameBillingCycle =
+      productPaymentCycleOld === baseDetails.productPaymentCycleNew;
+
+    // get next invoice details
+    const nextInvoice = await this.previewInvoiceBySubscriptionId({
+      subscriptionId: subscription.id,
+    });
+
+    const { total: nextInvoiceTotal, currency: nextInvoiceCurrency } =
+      nextInvoice;
+
+    // if same cycle, return next invoice lines total + tax
+    // else return invoice total
+    const nextInvoiceLines = nextInvoice.lines.data[1];
```

Comment:
What invoice line item is this referring to? Is the order in `lines` guaranteed, so that the required line item is always the 2nd entry? What if there are more than two line items?

Reply:
This one was tough. I wanted to get the list price and tax of the upgraded product. The first one shows the upgrade change itself since the current invoice does not include the proration. The second item shows the new total and the third shows the subsequent one (although I could be reading it incorrectly but that's how it appeared to me from the debugger and docs).

I am completely open to a simpler and more failproof solution. In the end we just want the new product total of list price + tax without any additional information.

Reply:
Looking at [the Stripe docs](https://stripe.com/docs/api/invoices/object#invoice_object-lines), it seems the invoice line_items are sorted, but the positions aren't guaranteed.

> The individual line items that make up the invoice. lines is sorted as follows: (1) pending invoice items (including prorations) in reverse chronological order, (2) subscription items in reverse chronological order, and (3) invoice items added after invoice creation in chronological order.

It is possible for there to be multiple invoice line items, in addition to the one's created by the upgrade. A possible solution to this could be to sum up the values of the the relevant invoice line items. (An example of this is the code that was removed by this PR)

Reply:
Tested this theory and I can absolutely confirm that restoring that code sums up all the amounts sans tax for the invoice, resulting in the **proration amount w/o tax**, and not the **list price w/ tax** as desired.

After reviewing [the Stripe docs on prorations](https://stripe.com/docs/billing/subscriptions/prorations), I was able to modify the code such that it targets the specific line item needed (the list price and tax on the new plan that matches what is displayed on checkout), regardless of position. That was a good catch. Thank you!